### PR TITLE
add keys & per-key products for apps

### PR DESCRIPTION
**Pull request by IanWillard**
Adding the ability to specify specific keys and associate them to specific products within a given app.  The intent is to allow an apigee server to be cloned such that clients are not aware of the change (and do not need to be re-released with new keys), or to allow for multiple products to be associated to a single key within a given app.

### CODE REVIEW

**Comment by msadasivam** [comment]
This is cool @IanWillard let me take a look.

**Comment by msadasivam** [comment]
@IanWillard I took a look at the changes. The part that makes Apps have multiple credentials and corresponding API products is useful (infact somebody is asking for it) but this change will allow/suggest developers checkin client id and secret to their git repos.

Migrating keys and secrets fits well with the migration tool. Is there a way to discourage git repos having id/secret for apps ?

**Comment by msadasivam** [comment]
I have been thinking about it, a way to achieve this is to use a separate encryption key for app key/secret or any sensitive data. This way what is stored in SCM will be encrypted and the plugin will decrypt it before deploying to Edge. the new option can be accepted as follows:
-Dapigee.config.securekey=<key>
Using a pbkdf2 function we will generate a key to decrypt the entry.

Correspondingly we we can enhance the migration tool to generated encrypted format. Let me know your thoughts.
@IanWillard @ssvaidyanathan

File: samples/EdgeConfig/edge.json [hunk 1]
```diff
@@ -148,8 +148,12 @@
             "bill@unesco.com": [
                 {
                     "name": "oneapp",
-                    "apiProducts": [
-                        "weatherProduct"
+                    "keys": [
```
Comment (ssvaidyanathan) [inline samples/EdgeConfig/edge.json]: not a valid JSON

File: samples/EdgeConfig/edge.json [hunk 2]
```diff
@@ -148,8 +148,12 @@
...
+                        "consumerKey": "i8wVEIbuLtedoy4gJE3IukSq2Gww0trT",
+                        "consumerSecret": "x8tFPqcmEAXm8ewG",
+                        "apiProducts": [
+                            "weatherProduct"
+                        ],
```
Comment (ssvaidyanathan) [inline samples/EdgeConfig/edge.json]: invalid json, remove ","

**Comment by msadasivam** [comment]
@IanWillard  can you please complete this agreement for us to pull your changes.
https://cla.developers.google.com/

**Comment by IanWillard** [comment]
I completed the agreement, not sure if you get notified.

**Comment by msadasivam** [comment]
Thanks @IanWillard.

Recently we added a feature to export newly registered app keys. This can be used to run tests in CI/CD pipeline https://github.com/apigee/apigee-config-maven-plugin/releases/tag/v1.1.4

Storing key/secret in source control even in encrypted form has been identified as an anti-pattern, so we are not looking to support key/secret in config files. If you need a way to migrate key/secret please leverage the migration tool.
https://github.com/apigeecs/apigee-migrate-tool

The ability to create multiple credentials in an app by including credential specific API Products is a necessary feature. Can you please remove the key/secret handling parts and send in a new pull request.

**Comment by kurtkanaskie** [comment]
I would use this enhancement as it supports the use case where external App keys (e.g. those created by another API management solution) are used in Edge. 

This can be done via management APIs currently.

And I think it should follow the pattern returned from the Mgmt API:
...
"name": "identity-app-cicd",
"credentials": [
        {
            "apiProducts": [
                {
                    "apiproduct": "identity-product-cicd",
                    "status": "approved"
                }
            ],
            "attributes": [],
            "consumerKey": "externalclientid",
            "consumerSecret": "supersecret",
            "issuedAt": 1507657002909,
            "scopes": [],
            "status": "approved"
        }
    ],
...

**Pull request closed without merging**